README.md
# IB/Extensions

---

__Documentation: [https://ib-ruby.github.io/ib-doc/](https://ib-ruby.github.io/ib-doc/)__  

__Questions, Contributions, Remarks: [Discussions are opened in ib-api](https://github.com/ib-ruby/ib-api/discussions)__


__Helpers and Macros that ease the usage of the TWS-API of Interactive Brokers__

To activate use
```
gem 'ib-extensions'
```
in the Gemfile and require the extensions as needed

## Include all
(except gateway)
```
require 'ib-api'
require 'ib/extensions'
```

## Verify
```
require 'ib-api'
require 'ib/extensions/verify'
```
Verifies a given contract, for details refer to the [documentation](https://ib-ruby.github.io/ib-doc/Verify_contracts.html )

## Market Price
```
require 'ib-api'
require 'ib/extensions/market-price'
```
Returns the most recent market-price of a given contract  ( [documentation](https://ib-ruby.github.io/ib-doc/market_price.html) )

## Historical Data (EOD)
```
require 'ib-api'
require 'ib/extensions/eod'

puts IB::Symbols.Index.estx.eod( duration: '10 d' )
```
Fetch historical data with just one line of code  ([documentation](https://ib-ruby.github.io/ib-doc/Historical_data.html) )

**NEW :: Contract.eod returns a PolarsDataframe **

## Order Prototypes
```
require 'ib-api'
require 'ib/order-prototypes'

order = IB::Limit.order size: 100, price: 10, action: :buy
order = IB::StopLimit.order size: 100, price: 10, stop_price: 9.5
```

then transmit the order through  the [place_order](https://ib-ruby.github.io/ib-doc/order_placement.html)  method of IB::Connection or  Account-based  preview, place, modify and cancel methods of [IB::Gateway](https://ib-ruby.github.io/ib-doc/order_placement.html).

More details in the [documentation](https://ib-ruby.github.io/ib-doc/order_prototypes.html) 


## Spread Prototypes

``` ruby
require 'ib-api'
require 'ib/spread-prototypes'
```

Compose most common spreads through

``` ruby
s = IB::Straddle.build from: IB::Symbols::Index.stoxx, 
                            strike: 4200, 
                            expiry: 202112 
                            
t = IB::Strangle.build from: IB::Symbols::Index.stoxx, 
                            c: 2400, p: 2200, 
                            expiry: 202103 

puts s.as_table
┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                      Straddle ESTX50(4200.0)[Dec 2021]                                      │
├────────┬────────┬────────────┬──────────┬──────────┬────────────┬───────────────┬───────┬────────┬──────────┤
│        │ symbol │ con_id     │ exchange │ expiry   │ multiplier │ trading-class │ right │ strike │ currency │
╞════════╪════════╪════════════╪══════════╪══════════╪════════════╪═══════════════╪═══════╪════════╪══════════╡
│ Spread │ ESTX50 │ -532066861 │   DTB    │          │     10     │               │       │        │   EUR    │
│ Option │ ESTX50 │  266033438 │   DTB    │ 20211217 │     10     │     OESX      │  put  │ 4200.0 │   EUR    │
│ Option │ ESTX50 │  266033423 │   DTB    │ 20211217 │     10     │     OESX      │ call  │ 4200.0 │   EUR    │
└────────┴────────┴────────────┴──────────┴──────────┴────────────┴───────────────┴───────┴────────┴──────────┘
 ```
and use the spreads like any other contract. ([documentation](https://ib-ruby.github.io/ib-doc/spreads.html))

## Gateway 
```
require 'ib-gateway'
```
IB::Gateway is an enhancement of IB::Connection. Upon initialization, it detects active accounts and stores them in thread safe arrays. 

``` ruby
g =  IB::Gateway.current
clients =  G.clients
puts client.first.portfolio_values.as_table
┌───────────┬─────────────────────────────────────────────┬─────┬──────────┬──────────┬───────────┬────────────┬──────────┐
│           │                                             │ pos │ entry    │ market   │ value     │ unrealized │ realized │
╞═══════════╪═════════════════════════════════════════════╪═════╪══════════╪══════════╪═══════════╪════════════╪══════════╡
│ Uxxxxxxx  │ Stock: BEPC USD NYSE                        │ 200 │   43.038 │   41.042 │    8208.4 │    -399.17 │          │
│ Uxxxxxxx  │ Option: CNHI 20210819 put 13.0 IDEM EUR     │  -2 │    0.386 │     0.01 │     -9.63 │     376.37 │          │
│ Uxxxxxxx  │ Stock: EQT SEK SFB                          │ 200 │    443.8 │  420.469 │  84093.85 │   -4666.19 │ -1369.02 │
│ Uxxxxxxx  │ Future: ESTX50 20210917 EUR                 │  -1 │   4145.3 │  4098.15 │  -40981.5 │      471.5 │          │
│ Uxxxxxxx  │ Option: ESTX50 20211217 call 4200.0 DTB EUR │  -4 │    97.85 │   99.398 │   -3975.9 │      -61.9 │          │
│ Uxxxxxxx  │ Option: IWM 20210903 call 230.0 AMEX USD    │  -4 │    1.943 │    0.124 │    -49.72 │     727.49 │          │

```
Details in the [documentation](https://ib-ruby.github.io/ib-doc/gateway.html)

Generally `puts IB::Model.as_table` provides a modern and convenient output for the console and notebooks.

```ruby
g.update_orders
puts g.clients.first.orders.as_table
┌──────────┬───────────┬─────────────────────────────────────────┬──────┬─────┬────────┬────────┬───────┬────────┐
│ account  │ status    │                                         │ Type │ tif │ action │ amount │ price │ id/fee │
╞══════════╪═══════════╪═════════════════════════════════════════╪══════╪═════╪════════╪════════╪═══════╪════════╡
│ U123456  │ Submitted │ Option: SLV 20210716 put 24.0 SMART USD │ LMT  │ GTC │ sell   │ 5.0    │ 0.98  │ 0      │
└──────────┴───────────┴─────────────────────────────────────────┴──────┴─────┴────────┴────────┴───────┴────────┘
puts g.clients.first.orders.contract.as_table
┌────────┬────────┬───────────┬──────────┬──────────┬────────────┬───────────────┬───────┬────────┬──────────┐
│        │ symbol │ con_id    │ exchange │ expiry   │ multiplier │ trading-class │ right │ strike │ currency │
╞════════╪════════╪═══════════╪══════════╪══════════╪════════════╪═══════════════╪═══════╪════════╪══════════╡
│ Option │ SLV    │ 456347029 │  SMART   │ 20210716 │    100     │      SLV      │  put  │   24.0 │   USD    │
└────────┴────────┴───────────┴──────────┴──────────┴────────────┴───────────────┴───────┴────────┴──────────┘
```
Its used in [Simple Monitor](https://github.com/ib-ruby/simple-monitor)


## Notebooks
**`IB-Ruby`** code can be executed in [iruby jupyter notebooks](https://github.com/SciRuby/iruby). A few scripts are included in [IB-Examples](https://github.com/ib-ruby/ib-examples). (*.ipynb- files)


```
## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/ib-ruby/ib-extensions. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[ib-ruby/ib-extensions/blob/master/CODE_OF_CONDUCT.md).


## Code of Conduct

Everyone interacting in the Ib::Extensions project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/ib-ruby/ib-extensions/blob/master/CODE_OF_CONDUCT.md).


changelog.md

* lib/gateway.rb
	* rename `active_accounts` to `clients`  
	  Clients are just tradeable `IB::Accounts`.  
 
  * rename ' for_active_accounts' to `account_data`
		Account-data provides a thread-safe access to contracts, portfolio- and account-values of a given account.

  * rename `simple_account_data_scan`  to `account_Data_scan`

  * connect: raises an error if connection is not possible because of using a not unique client_id

	* initialize::  tolerates arguments other then specified. 

	* made methods that never should call by users private:  
	  check_connection, initialize_alerts, initialize_managed_accounts, prepare_connection 
 
* lib/ib/gateway   
  
  Directory contains gateway related stuff

	* account-infos.rb
	* order-handling.rb

* lib/ib/alerts
   
	* included specific alert-definitions in this directory:
	  order-alerts, gateway-alerts

* moving model/spread.rb to `ib-api`  Gem

#### Preparation of a Gem-Release 

* Gateway#connect: initializing order-array by calling :RequestAllOrders after establishing the connection.

* Gateway#connect: Occasionally the request for AccountPositions/AllOrders fails. Then a reconnect is
                   appropriate. This is now implemented.

* Order.auto_adjust

     Auto Adjust implements a simple algorithm to ensure that an order is accepted
    
     It reads `contract_detail.min_tick`. 
    
     If min_tick < 0.01, the real tick-increments differ from the min_tick_value
    
     For J36 (jardines) min tick is 0.001, but the minimal increment is 0.005
     For Tui1 its the same, min_tick is 0.00001 , minimal increment is 0.00005
    
     Thus, for min-tick smaller then 0.01, the value is rounded to the next higher digit.
     
     | min-tick     |  round     |
     |--------------|------------|
     |   10         |   110      |
     |    1         |   111      |
     |    0.1       |   111.1    |
     |    0.001     |   111.11   |
     |    0.0001    |   111.11   |
     |    0.00001   |   111.111  |
     |--------------|------------|

* Client.place:   defaults are 
  *   auto_adjust is "true" 
  *   convert_size: true

  If a negative value for 'total amount' is used, then Order.action is set to :sell.   
  If a positive value is used, Order.action is kept 

* Client.place raises an `IB::Symbol Error` if the order is not submitted 

* Included Contract#included_in?( an IB::Account ) and Contract#portfolio_value( an IB::Account )

  Suppose, you want to check, if a complex Contract is included in the Portfolio

  ```ruby
  s = Straddle.build symbol: Symbols::Index.stoxx, strike=4200, expiry: 20211217
  puts s.included__in?(IB::Gateway.current.clients.first).to_human
  =>  "<Straddle ESTX50(4200.0)[Dec 2021]>" 
  puts s.portfolio_value( IB::Gateway.current.clients.first ).to_human
  => <PortfolioValue: Uxxxxxx Pos=-4 @ 225.158;Value=-9006.31;PNL=-1812.31 unrealized;<Option: ESTX50 20211217 put 4200.0 DTB EUR>
 => <PortfolioValue: Uxxxxxxx Pos=-4 @ 99.398;Value=-3975.9;PNL=-61.9 unrealized;<Option: ESTX50 20211217 call 4200.0 DTB EUR>

  ```
  Both methods are available for IB::Contracts.

* IB::Gateway.current.get_account_data  accepts only one parameter and is not persistent anymore
   The former second argument (watchlists: ) is defaulted by those initialized through e IB::Gateway.new  (IB::Gateway.current.active_watchlists)  
   The method now fetches the account- and portfoliodate, initialises Account#contracts, Account#portfolio_values, Account#account_values , Account#focusses and desubscribes from the TWS. (Before, the subscription to AccountData was permanently active).

* IB::Gateway.current.add_watchlist( a symbol ) -- Watchlists can added after initialisation

* New directory structure:  IB::Extensions::Modulename
* Extensions are modules which are included in IB::Contracts
* IB::Option.underlying fetches the contract as specified in contract-details of  IB::Option


lib/ib/models/order.rb

module IB

  class Order
    # Auto Adjust implements a simple algotithm to ensure that an order is accepted
    #
    # It reads `contract_detail.min_tick`. 
    #
    # If min_tick < 0.01, the real tick-increments differ fron the min_tick_value
    #
    # For J36 (jardines) min tick is 0.001, but the minimal increment is 0.005
    # For Tui1 its the samme, min_tick is 0.00001 , minimal increment ist 0.00005
    #
    # Thus, for min-tick smaller then 0.01, the value is rounded to the next higer digit.
    # 
    # | min-tick     |  round     |
    # |--------------|------------|
    # |   10         |   110      |
    # |    1         |   111      |
    # |    0.1       |   111.1    |
    # |    0.001     |   111.11   |
    # |    0.0001    |   111.11   |
    # |    0.00001   |   111.111  |
    # |--------------|------------|
    #
    def auto_adjust
      # lambda to perform the calculation
      adjust_price = ->(a,b) do
        count = -Math.log10(b).round.to_i
        count = count -1 if count > 2
        a.round count

      end


      error "No Contract provided to Auto adjust" unless contract.is_a? IB::Contract

      unless contract.is_a? IB::Bag
        # ensure that contract_details are present

        min_tick = contract.verify.first.contract_detail.min_tick
        # there are two attributes to consider: limit_price and aux_price
        # limit_price +  aux_price may be nil or an empty string. Then ".to_f.zero?" becomes true 
        self.limit_price= adjust_price.call(limit_price.to_f, min_tick) unless limit_price.to_f.zero?
        self.aux_price= adjust_price.call(aux_price.to_f, min_tick) unless aux_price.to_f.zero?
      end
    end
  end  # class Order
end  # module
